### PR TITLE
new persistence api

### PR DESCRIPTION
**Pull request by raskolnikoov**
Rewrite of persistence api and added ravendb persistence plugin. Maybe ravendb/couchebase should be moved outside ofta the core framework into its own repos?

### CODE REVIEW

**Comment by raskolnikoov** [comment]
Im not 100% sure I got the versioning right. Byt maybe that is not set in the projects.

**Comment by rogeralsing** [comment]
Can you describe the changes to start a discussion about it?
How is snapshotting handled now?

**Comment by raskolnikoov** [comment]
I have removed the "GetSnapshotInterval" method and property from Persistence. I also removed the bool property "Recovering". I think snapshots should be triggered by the actor itself instead when snapshots are needed.

I've introduced the following messages to be send to the actor on recovery/persisting:

- RecoveryStarted
- RecoveryCompleted
- RecoverSnapshot *contains persisted snapshot object
- RecoverEvent *contains persisted event object
- PersistedEvent *contains the persisted event object

**Comment by raskolnikoov** [comment]
"IProviderState" has been expanded with methods to handle "DeleteEventsAsync" and "DeleteSnapshotsAsync" so the actor could clean up stored events/snapshots that are not needed anymore.

**Comment by cpx86** [comment]
Letting the actor control snapshotting makes sense IMO. It is more related to the needs of the actor than the needs of the persistence.

About versioning, don't worry. It's currently set by the build script and we only bump it after merging to master.

**Comment by raskolnikoov** [comment]
Anyone time ro review my changes?

**Comment by rogeralsing** [comment]
The change looks good to me.
cc @cpx86

File: src/Proto.Persistence/InMemoryProviderState.cs [hunk 1]
```diff
@@ -60,5 +51,15 @@ public Task PersistSnapshotAsync(string actorName, ulong eventIndex, object snap
             _snapshots[actorName] = Tuple.Create((object) snapshot, eventIndex);
             return Task.FromResult(0);
         }
+
+        public Task DeleteEventsAsync(string actorName, ulong fromEventIndex)
+        {
+            throw new NotImplementedException();
```
Comment (cpx86) [inline src/Proto.Persistence/InMemoryProviderState.cs]: We should just return `Task.FromResult(0)` here.

File: src/Proto.Persistence/InMemoryProviderState.cs [hunk 2]
```diff
@@ -60,5 +51,15 @@ public Task PersistSnapshotAsync(string actorName, ulong eventIndex, object snap
...
         }
+
+        public Task DeleteEventsAsync(string actorName, ulong fromEventIndex)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task DeleteSnapshotsAsync(string actorName, ulong fromEventIndex)
+        {
+            throw new NotImplementedException();
```
Comment (cpx86) [inline src/Proto.Persistence/InMemoryProviderState.cs]: These should probably be implemented as well. If we want to keep this class - I'm not sure if it's needed for anything except testing (in which case it should be moved)

**Comment by cpx86** [comment]
Looks good! I like the introduction of the recovery messages, so the actor can handle those differently if necessary :)

**Comment by cpx86** [comment]
Just fix the merge conflict and I'll merge 👍

File: src/Proto.Persistence/InMemoryProviderState.cs (hunk 2 above)
Comment (raskolnikoov) [inline src/Proto.Persistence/InMemoryProviderState.cs]: I think we should remove InMemoryProvider. Dont see the use case for it really. If we need a default provider maybe we should persist om disk in a File.

Comment (cpx86) [inline src/Proto.Persistence/InMemoryProviderState.cs]: Do it then :)

**Comment by rogeralsing** [comment]
Merged, thanks!

**Pull request merged**